watcher/src/watcher.rs

```diff
@@ -29,7 +28,16 @@ use url::Url;
 
 /// Watches multiple consensus validators and collects block signatures.
 pub struct Watcher {
-    transactions_fetcher: Arc<ReqwestTransactionsFetcher>,
+    /// A transaction fetcher per watched URL.
+    // The reason we keep a transaction fetcher per url has to do with the pre-fetching and caching
+    // mechanism implemented in ReqwestTransactionsFetcher: ReqwestTransactionsFetcher will try and
+    // fetch the large merged-blocks and then try to hand out results from its cache. However, it
+    // is oblivious as to which URL got used to cache the blocks. This behavior is suitable for
+    // LedgerSyncService but in the watcher we want to ensure the blocks are fetched from a
+    // specific URL. If we want to use the cache mechanism (which we do, since it cuts down sync
+    // time significantlly) then the workaround is to have a ReqwestTransactionsFetcher that only
+    // has a single source URL.
+    transactions_fetcher_by_url: Arc<HashMap<Url, ReqwestTransactionsFetcher>>,
```

Comment:
I think I'm not fully clear on the need to separate the cache by url. Is it on account of the watching wanting to grab the same block from several sources but still wanting to utilize the cache if you're requesting the same block from the same source?

Reply:
The cache is a bit non-trivial - the name cache might not be perfectly suitable. While you can get individual blocks from S3, you can also get what is called "merged blocks" which are files that contain multiple blocks. So if for example I ask for block 1000 and this is available as a merged block that covers blocks [1000-1999], the fetcher will fetch that and keep those blocks in the "cache" so that when the next request for block 1001 comes it does not hit S3 but instead is served from this cache.
`ReqwestTransactionsFetcher` supports multiple S3 source URLs that it round-robins (IIRC) from, which is good for when mobilecoind or consensus wants a block but don't care where it comes from. For the watcher we actually do care (we want to get the data from each of the available source URLs), hence a map that maps a source url to a `ReqwestTransactionsFetcher` that is configured to only fetch from that one particular URL.

Reply:
Thanks for the detail.